src/inlinestrings.jl

```diff
@@ -0,0 +1,770 @@
+                            InlineString3,
+                            InlineString7,
+                            InlineString15,
+                            InlineString31,
+                            InlineString63,
+                            InlineString127,
+                            InlineString255}
+
+function Base.promote_type(::Type{T}, ::Type{S}) where {T <: InlineString, S <: InlineString}
+    T === InlineString1 && return S
+    S === InlineString1 && return T
+    T === InlineString3 && return S
+    S === InlineString3 && return T
+    T === InlineString7 && return S
+    S === InlineString7 && return T
+    T === InlineString15 && return S
+    S === InlineString15 && return T
+    T === InlineString31 && return S
+    S === InlineString31 && return T
+    T === InlineString63 && return S
+    S === InlineString63 && return T
+    T === InlineString127 && return S
+    S === InlineString127 && return T
+    return InlineString255
+end
+
+Base.ncodeunits(::InlineString1) = 1
+Base.ncodeunits(x::InlineString) = Int(Base.trunc_int(UInt8, x))
+Base.codeunit(::InlineString) = UInt8
+
+Base.@propagate_inbounds function Base.codeunit(x::T, i::Int) where {T <: InlineString}
+    @boundscheck checkbounds(Bool, x, i) || throw(BoundsError(x, i))
+    if T === InlineString1
+        return Base.bitcast(UInt8, x)
+    else
+        return Base.trunc_int(UInt8, Base.lshr_int(x, 8 * (sizeof(T) - i)))
+    end
+end
+
+function Base.String(x::T) where {T <: InlineString}
+    len = ncodeunits(x)
+    out = Base._string_n(len)
+    if T === InlineString1
+        GC.@preserve out unsafe_store!(pointer(out), codeunit(x, 1))
+        return out
+    end
+    ref = Ref{T}(_bswap(x))
+    GC.@preserve ref out begin
+        ptr = convert(Ptr{UInt8}, Base.unsafe_convert(Ptr{T}, ref))
+        unsafe_copyto!(pointer(out), ptr, len)
+    end
+    return out
+end
+
+function Base.Symbol(x::T) where {T <: InlineString}
+    ref = Ref{T}(_bswap(x))
+    return ccall(:jl_symbol_n, Ref{Symbol},
+        (Ref{T}, Int), ref, sizeof(x))
+end
+
+# add a codeunit to end of string method
+function addcodeunit(x::T, b::UInt8) where {T <: InlineString}
+    if T === InlineString1
+        return x, false
+    end
+    len = ncodeunits(x)
+    (len + 1) == sizeof(T) && return x, true
+    x = Base.or_int(x, Base.shl_int(Base.zext_int(T, b), 8 * (sizeof(T) - len - 1)))
+    return Base.add_int(x, Base.zext_int(T, 0x01)), false
+end
+
+# from String
+InlineString1(byte::UInt8=0x00) = Base.bitcast(InlineString1, byte)
+(::Type{T})() where {T <: InlineString} = Base.zext_int(T, 0x00)
+
+function (::Type{T})(x::AbstractString) where {T <: InlineString}
+    if T === InlineString1
+        sizeof(x) == 1 || stringtoolong(T, sizeof(x))
+        return Base.bitcast(InlineString1, codeunit(x, 1))
+    elseif typeof(x) === String && sizeof(T) <= sizeof(UInt)
+        len = sizeof(x)
+        len < sizeof(T) || stringtoolong(T, len)
```

Comment:
just to be clear - the condition here is acute because you need to leave the last byte unused?

Reply:
Correct; the last byte (in reality, it's the first byte or lowest byte) is used to track the length of how many bytes are actually used in the inline string